src/main/java/com/google/opensesame/servlets/ProjectServlet.java

```diff
@@ -7,54 +7,117 @@
 import com.google.opensesame.projects.ProjectEntity;
 import com.google.opensesame.util.ErrorResponse;
 import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
-@WebServlet("/project")
+@WebServlet("/projects")
 public class ProjectServlet extends HttpServlet {
 
   private final Gson gson = new Gson();
 
   private static final String PROJECT_ID_PARAM = "projectId";
+  private static final String FULL_DATA_PARAM = "fullData";
+  private static final String PREVIEW_DATA_PARAM = "previewData";
 
   /**
-   * @return Returns the request parameter associated with the inputted name, or returns the default
-   *     value if the specified parameter is not defined.
+   * Returns a boolean parameter from a request. If the parameter does not exist, its value is
+   * assumed to be false, but if the value does exist and it is not a valid boolean it returns null.
+   *
+   * @param request The request to get the parameter from.
+   * @param parameter The name of the boolean parameter.
+   * @return Returns the value of the parameter, the assumed value, or null if the value is not a
+   *     valid boolean.
    */
-  public static String getParameter(HttpServletRequest request, String name, String defaultValue) {
-    String value = request.getParameter(name);
-    if (value == null) {
-      return defaultValue;
+  private static Boolean booleanFromRequest(HttpServletRequest request, String parameter) {
+    String parameterValue = request.getParameter(parameter);
+    if (parameterValue == null) {
+      return Boolean.valueOf(false);
+    }
+
+    if (parameterValue.equalsIgnoreCase("true") || parameterValue.equalsIgnoreCase("false")) {
+      return Boolean.valueOf(parameterValue);
+    } else {
+      return null;
     }
-    return value;
+  }
+
+  /**
+   * Creates a collection of ProjectData from a collection of ProjectEntities.
+   *
+   * @param projectEntities
+   * @param getFullData
+   * @param getPreviewData
+   * @return Returns the created collection of ProjectData.
+   * @throws IOException Throws if there was an error retrieving the GitHub repository associated
+   *     with a ProjectEntity.
+   */
+  private static Collection<ProjectData> createProjectData(
+      Collection<ProjectEntity> projectEntities, boolean getFullData, boolean getPreviewData)
+      throws IOException {
+    List<ProjectData> projectData = new ArrayList<ProjectData>();
+    for (ProjectEntity projectEntity : projectEntities) {
+      ProjectData newProjectData = new ProjectData(projectEntity);
+      // TODO(Richie): Is there a better way to do this than an if-else block?
```

Comment:
That's a good point; I was more thinking about how I could design it such that the if-else block could be avoided altogether, but I'll definitely take another look at this code and see if the constructor idea fits better.

Reply:
My 2¢: the bools are not good, I agree.

How about this: ProjectData just has separate getters for description, topicTags, etc. If we want to do caching there, we can have private class variables initialized to null and only get the data from Github when they're null.

Instead of passing booleans to one function, we can have `createProjectPreviewData`, `createProjectFullData`, and `createProjectData` functions that just get the relevant fields.

And actually looking at the boolean logic below I feel it'd be simpler to just have different servlets/endpoints for each of the above methods.

Reply:
> How about this: ProjectData just has separate getters for description, topicTags, etc. If we want to do caching there, we can have private class variables initialized to null and only get the data from Github when they're null.

Keep in mind this object is being prepared for serialization and not much else, so the main intention is to assign the values to the class variables before the object is serialized to JSON.

> Instead of passing booleans to one function, we can have createProjectPreviewData, createProjectFullData, and createProjectData functions that just get the relevant fields.

This sounds very similar to what has already been implemented, so please let me know if I'm not understanding it completely.

> And actually looking at the boolean logic below I feel it'd be simpler to just have different servlets/endpoints for each of the above methods.

This would definitely remove the need for an if-else block anywhere in the code. I'm thinking abstracting the `ProjectEntity` query system to avoid repeat and using the different endpoints to implicitly determine what data is needed could work!

Reply:
Unless there are any objections I'm definitely going to move forward with the different endpoints idea, thanks Bill and Sami!

Reply:
Ah yeah actually I wasn't sure about the purpose of this class, even though I read the comment in ProjectData, line 15. Maybe just say "This is a read-only class is for serializing data from Datastore to JSON" - certain buzzwords like "read-only" convey a lot. 

So yeah the logic can be even simpler then, no null-checking, but I still think there should be getters.

IMO it's not a good assumption to make that e.g. `withFullData` will always call `withPreviewData` which will always call `withName`, I feel that could easily change in the future on a whim depending on what we want to throw on the pages. 

I think it's best to be able to just specify what you want. Currently what we have is similar to an inheritance-based approach which requires a lot of layers of looking at what data this function gets, which calls this function which calls this function, etc. and isn't very flexible. Maybe that's the premature optimization that was referred to?

Reply:
Yeah that makes a lot of sense when you say it that way. I'll revise based on those recommendations, thanks so much!